### ISSUE TITLE
SharpMap WMTS / TMS Server implementaion

### ISSUE THREAD

**Issue by tkmawarire** [SharpMap/SharpMap#93]
Can anyone help me with the implementation of WMTS / TMS Server in SharpMap. Been trying out a lot from diff sources but i can seen to come up with a working solution.Below is a handler i am using, it's just drawing the boxes instead of data from the database. 

```
    /// <summary>
    /// Summary description for WMTS
    /// </summary>
    public class WMTS : IHttpHandler
    {
        /// <summary>
        /// Defines the projection
        /// </summary>
        private readonly ICoordinateTransformation projection = ProjUtil.ToPseudoMercator();

        /// <summary>
        /// The ProcessRequest
        /// wmts?SERVICE=WMTS&REQUEST=GetTile&VERSION=1.0.0&LAYER=layer_id&STYLE=default&TILEMATRIXSET=matrix_id&TILEMATRIX=3&TILEROW=2&TILECOL=0&FORMAT=image%2Fjpeg
        /// </summary>
        /// <param name="context">The <see cref="HttpContext"/></param>
        public void ProcessRequest(HttpContext context)
        {
            string layer = context.Request.Params["LAYER"];
            int tilematrix = int.Parse(context.Request.Params["TILEMATRIX"]);
            int tilerow = int.Parse(context.Request.Params["TILEROW"]);
            int tilecol = int.Parse(context.Request.Params["TILECOL"]);
            string service = context.Request.Params["SERVICE"];
            string request = context.Request.Params["REQUEST"];
            string version = context.Request.Params["VERSION"];
            string style = context.Request.Params["STYLE"];
            string tilematrixset = context.Request.Params["TILEMATRIXSET"];
            string format = context.Request.Params["FORMAT"];

            if (String.IsNullOrEmpty(layer))
                throw new ArgumentNullException("layer");

            Map map = Map(layer, tilecol, tilerow, tilematrix);

            var map_image = map.GetMap();

            //using (var memory_stream = new MemoryStream())
            //{
            //    map_image.Save(memory_stream, ImageFormat.Png);

            //    var wms = memory_stream.ToArray();

            //    WriteResponseInChunks(wms, context);
            //}
            byte[] buffer;
            using (var ms = new MemoryStream())
            {
                map_image.Save(ms, ImageFormat.Png);
                map_image.Dispose();
                buffer = ms.ToArray();
            }
            WriteResponseInChunks(buffer, context);

        }

        public static ImageCodecInfo GetEncoderInfo(String mimeType)
        {
            foreach (var encoder in ImageCodecInfo.GetImageEncoders())
                if (encoder.MimeType == mimeType)
                    return encoder;
            return null;
        }
        /// <summary>
        /// The GetMap
        /// </summary>
        /// <returns>The <see cref="SharpMap.Map"/></returns>
        protected Map Map(string layer, int x, int y, int z)
        {
            Envelope bbox = GetBoundingBoxInLatLngWithMargin(x, y, z);

            return DatabaseUtil.SqlServer(ConnectionString(), Layers().FindAll(lyr => lyr.Table.Equals(layer)), new Size(1, 1), bbox, "id");
        }

        /// <summary>
        /// The Layers
        /// </summary>
        /// <returns>The <see cref="List{VectorLayerModel}"/></returns>
        private List<VectorLayerModel> Layers()
        {

            VectorLayerModel standsLayerModel = new VectorLayerModel()
            {
                Table = "cadastre",
                Style = new VectorStyle { Line = new Pen(Color.DarkGray, 2) }
            };
            VectorLayerModel roadsLayerModel = new VectorLayerModel()
            {
                Table = "townships",
                Style = new VectorStyle { Line = new Pen(Color.DarkRed, 2.5f) }
            };
            VectorLayerModel pipeLayerModel = new VectorLayerModel()
            {
                Table = "provinces",
                Style = new VectorStyle { Line = new Pen(Color.DarkBlue, 1.5f) }
            };

            return new List<VectorLayerModel>() { standsLayerModel, roadsLayerModel, pipeLayerModel };
        }

        /// <summary>
        /// The ConnectionString
        /// </summary>
        /// <returns>The <see cref="string"/></returns>
        private string ConnectionString()
        {
            return "Data Source=******;Initial Catalog=GCCIGO_V2;Integrated Security=SSPI;";
        }

        /// <summary>
        /// The GetBoundingBoxInLatLngWithMargin
        /// </summary>
        /// <param name="tileX">The <see cref="int"/></param>
        /// <param name="tileY">The <see cref="int"/></param>
        /// <param name="zoom">The <see cref="int"/></param>
        /// <returns>The <see cref="Envelope"/></returns>
        private Envelope GetBoundingBoxInLatLngWithMargin(int tileX, int tileY, int zoom)
        {
            Point px1 = new Point((tileX * 256), (tileY * 256));
            Point px2 = new Point(((tileX + 1) * 256), ((tileY + 1) * 256));

            PointF ll1 = TileSystemHelper.PixelXYToLatLong(px1, zoom);
            PointF ll2 = TileSystemHelper.PixelXYToLatLong(px2, zoom);

            double[] prj1 = projection.MathTransform.Transform(new double[] { ll1.X, ll1.Y });
            double[] prj2 = projection.MathTransform.Transform(new double[] { ll2.X, ll2.Y });

            Envelope bbox = new Envelope();
            bbox.ExpandToInclude(prj1[0], prj1[1]);
            bbox.ExpandToInclude(prj2[0], prj2[1]);
            return bbox;
        }

        /// <summary>
        /// The size of the chunks written to response.
        /// </summary>
        private const int ChunkSize = 2 * 8192;

        /// <summary>
        /// Method to write an array of bytes in chunks to a http response
        /// </summary>
        /// <remarks>
        /// The code was adopted from http://support.microsoft.com/kb/812406/en-us
        /// </remarks>
        /// <param name="buffer">The array of bytes</param>
        /// <param name="context">The response</param>
        private static void WriteResponseInChunks(byte[] buffer, HttpContext context)
        {
            try
            {
                bool _continue;

                context.Response.ClearContent();

                context.Response.ContentType = "image/png";

                using (var ms = new MemoryStream(buffer))
                {
                    var dataToRead = buffer.Length;

                    while (dataToRead > 0)
                    {
                        if (context.Response.IsClientConnected)
                        {
                            {
                                var tmpBuffer = new byte[ChunkSize];
                                var length = ms.Read(tmpBuffer, 0, tmpBuffer.Length);
                                context.Response.OutputStream.Write(tmpBuffer, 0, length);
                                context.Response.Flush();
                                dataToRead -= length;
                            }
                        }
                        else
                        {
                            dataToRead = -1;
                        }
                    }
                    _continue = dataToRead > 0;

                }
            }
            catch (Exception ex)
            {
                context.Response.ClearContent();
                context.Response.ContentType = "text/plain";
                context.Response.Write(string.Format("Error     : {0}", ex.Message));
                context.Response.Write(string.Format("Source    : {0}", ex.Message));
                context.Response.Write(string.Format("StackTrace: {0}", ex.StackTrace));
            }
            finally
            {
                context.Response.End();
            }
        }

        /// <summary>
        /// Gets a value indicating whether IsReusable
        /// </summary>
        public bool IsReusable
        {
            get
            {
                return true;
            }
        }
    }
```

**Comment by FObermaier** [comment]
If I understand you correctly, then you are getting a result, but the result is an empty map.
If so, could you 
* show us the code to generate the `Map` (`DatabaseUtil.SqlServer(...)`)?
* tell us what kind of spatial type you use (SqlGeometry/SqlGeography) and in which spatial reference system the data is stored.
* tell us whether or not you apply some coordinate transformation on the data.

**Comment by tkmawarire** [comment]
Sure, am using SQLGeometry in WGS84 (4326). I didn't apply any transformations as that kinda gave me an error when i was working on WMS implementation. The WMS is working fine without the coordinate transformation so i thought i would skip it.  Anyway below is the code generating the `Map`:

```
public static Map SqlServer(string conn, List<VectorLayerModel> layers, Size map_size,Envelope bbox, string id_column = "ID")
        {
            Map map = new Map(map_size);

            foreach (var layer in layers)
            {
                VectorLayer lyr = CreateSqlServerLayer(conn, layer.Table, layer.Style, id_column);
                lyr.IsQueryEnabled = true;
                lyr.Enabled = true;
                if (bbox != null)
                {
                    var geometries = lyr.DataSource.GetGeometriesInView(bbox);
                    lyr.DataSource = new GeometryFeatureProvider(geometries);
                    
                }
                map.Layers.Add(lyr);
            }

            return map;
        }
```
The `CreateSqlServerLayer`:

```
        /// <summary>
        /// Create a SharpMap Vector Layer
        /// </summary>
        /// <param name="connstr">The <see cref="string"/></param>
        /// <param name="name">The <see cref="string"/></param>
        /// <param name="style">The <see cref="VectorStyle"/></param>
        /// <returns>The <see cref="VectorLayer"/></returns>
        private static VectorLayer CreateSqlServerLayer(string connstr, string name, VectorStyle style, string idColumnName)
        {
            if (connstr == null)
                throw new ArgumentNullException("connstr");
            if (name == null)
                throw new ArgumentNullException("name");


            SqlServer2008 source = new SqlServer2008(connstr, name, "geom", idColumnName) { ValidateGeometries = true };

            ICoordinateTransformation transformation = ProjUtil.ToPseudoMercator();

            VectorLayer item = new VectorLayer(name, source)
            {
                SRID = 4326,
                TargetSRID = 4326,
                //TargetSRID = 900913,
                //CoordinateTransformation = transformation,
                Style = style,
                SmoothingMode = SmoothingMode.AntiAlias,
            };

            return item;
   }
```
And finally the `VectorLayerModel`:

```
    /// <summary>
    /// Defines the <see cref="VectorLayerModel" />
    /// </summary>
    public class VectorLayerModel
    {
        /// <summary>
        /// Gets or Sets the SQL Server Column
        /// </summary>
        public string Table { get; set; }

        /// <summary>
        /// Gets or sets the Vector Layer Style
        /// </summary>
        public VectorStyle Style { get; set; }
    }
```
Am just using this as a dictionary for my `VectorLayer` defination i.e. the style & the server tablename

**Comment by tkmawarire** [comment]
Yah, i didn't have thats that in my code. I am still getting just red square lines and no data. 
```
    /// <summary>
    /// Summary description for WMTS
    /// </summary>
    public class WMTS : IHttpHandler
    {
        /// <summary>
        /// Defines the projection
        /// </summary>
        private readonly ICoordinateTransformation projection = ProjUtil.ToPseudoMercator();

        /// <summary>
        /// The ProcessRequest
        /// wmts?SERVICE=WMTS&REQUEST=GetTile&VERSION=1.0.0&LAYER=layer_id&STYLE=default&TILEMATRIXSET=matrix_id&TILEMATRIX=3&TILEROW=2&TILECOL=0&FORMAT=image%2Fjpeg
        /// </summary>
        /// <param name="context">The <see cref="HttpContext"/></param>
        public void ProcessRequest(HttpContext context)
        {
            string layer = context.Request.Params["LAYER"];
            int tilematrix = int.Parse(context.Request.Params["TILEMATRIX"]);
            int tilerow = int.Parse(context.Request.Params["TILEROW"]);
            int tilecol = int.Parse(context.Request.Params["TILECOL"]);
            string service = context.Request.Params["SERVICE"];
            string request = context.Request.Params["REQUEST"];
            string version = context.Request.Params["VERSION"];
            string style = context.Request.Params["STYLE"];
            string tilematrixset = context.Request.Params["TILEMATRIXSET"];
            string format = context.Request.Params["FORMAT"];

            if (String.IsNullOrEmpty(layer))
                throw new ArgumentNullException("layer");

            Envelope bbox = GetBoundingBoxInLatLngWithMargin(tilecol, tilerow, tilematrix);

            Map map = Map(layer, bbox);

            map.ZoomToBox(bbox);

            var map_image = map.GetMap();

            using (var memory_stream = new MemoryStream())
            {
                map_image.Save(memory_stream, ImageFormat.Png);

                var wms = memory_stream.ToArray();

                WriteResponseInChunks(wms, context);
            }

        }

        public static ImageCodecInfo GetEncoderInfo(String mimeType)
        {
            foreach (var encoder in ImageCodecInfo.GetImageEncoders())
                if (encoder.MimeType == mimeType)
                    return encoder;
            return null;
        }
        /// <summary>
        /// The GetMap
        /// </summary>
        /// <returns>The <see cref="SharpMap.Map"/></returns>
        protected Map Map(string layer, Envelope bbox)
        {
            
            return DatabaseUtil.SqlServer(ConnectionString(), Layers().FindAll(lyr => lyr.Table.Equals(layer)), new Size(1, 1), bbox, "id");
        }

        /// <summary>
        /// The Layers
        /// </summary>
        /// <returns>The <see cref="List{VectorLayerModel}"/></returns>
        private List<VectorLayerModel> Layers()
        {

            VectorLayerModel standsLayerModel = new VectorLayerModel()
            {
                Table = "cadastre",
                Style = new VectorStyle { Line = new Pen(Color.DarkGray, 2) }
            };
            VectorLayerModel roadsLayerModel = new VectorLayerModel()
            {
                Table = "townships",
                Style = new VectorStyle { Line = new Pen(Color.DarkRed, 2.5f) }
            };
            VectorLayerModel pipeLayerModel = new VectorLayerModel()
            {
                Table = "provinces",
                Style = new VectorStyle { Line = new Pen(Color.DarkBlue, 1.5f) }
            };

            return new List<VectorLayerModel>() { standsLayerModel, roadsLayerModel, pipeLayerModel };
        }

        /// <summary>
        /// The ConnectionString
        /// </summary>
        /// <returns>The <see cref="string"/></returns>
        private string ConnectionString()
        {
            return "Data Source=*****;Initial Catalog=*******;Integrated Security=SSPI;";
        }

        /// <summary>
        /// The GetBoundingBoxInLatLngWithMargin
        /// </summary>
        /// <param name="tileX">The <see cref="int"/></param>
        /// <param name="tileY">The <see cref="int"/></param>
        /// <param name="zoom">The <see cref="int"/></param>
        /// <returns>The <see cref="Envelope"/></returns>
        private Envelope GetBoundingBoxInLatLngWithMargin(int tileX, int tileY, int zoom)
        {
            Point px1 = new Point((tileX * 256), (tileY * 256));
            Point px2 = new Point(((tileX + 1) * 256), ((tileY + 1) * 256));

            PointF ll1 = TileSystemHelper.PixelXYToLatLong(px1, zoom);
            PointF ll2 = TileSystemHelper.PixelXYToLatLong(px2, zoom);

            double[] prj1 = projection.MathTransform.Transform(new double[] { ll1.X, ll1.Y });
            double[] prj2 = projection.MathTransform.Transform(new double[] { ll2.X, ll2.Y });

            Envelope bbox = new Envelope();
            bbox.ExpandToInclude(prj1[0], prj1[1]);
            bbox.ExpandToInclude(prj2[0], prj2[1]);
            return bbox;
        }

        /// <summary>
        /// The size of the chunks written to response.
        /// </summary>
        private const int ChunkSize = 2 * 8192;

        /// <summary>
        /// Method to write an array of bytes in chunks to a http response
        /// </summary>
        /// <remarks>
        /// The code was adopted from http://support.microsoft.com/kb/812406/en-us
        /// </remarks>
        /// <param name="buffer">The array of bytes</param>
        /// <param name="context">The response</param>
        private static void WriteResponseInChunks(byte[] buffer, HttpContext context)
        {
            try
            {
                bool _continue;

                context.Response.ClearContent();

                context.Response.ContentType = "image/png";

                using (var ms = new MemoryStream(buffer))
                {
                    var dataToRead = buffer.Length;

                    while (dataToRead > 0)
                    {
                        if (context.Response.IsClientConnected)
                        {
                            {
                                var tmpBuffer = new byte[ChunkSize];
                                var length = ms.Read(tmpBuffer, 0, tmpBuffer.Length);
                                context.Response.OutputStream.Write(tmpBuffer, 0, length);
                                context.Response.Flush();
                                dataToRead -= length;
                            }
                        }
                        else
                        {
                            dataToRead = -1;
                        }
                    }
                    _continue = dataToRead > 0;

                }
            }
            catch (Exception ex)
            {
                context.Response.ClearContent();
                context.Response.ContentType = "text/plain";
                context.Response.Write(string.Format("Error     : {0}", ex.Message));
                context.Response.Write(string.Format("Source    : {0}", ex.Message));
                context.Response.Write(string.Format("StackTrace: {0}", ex.StackTrace));
            }
            finally
            {
                context.Response.End();
            }
        }

        /// <summary>
        /// Gets a value indicating whether IsReusable
        /// </summary>
        public bool IsReusable
        {
            get
            {
                return true;
            }
        }
    }
```

**Comment by tkmawarire** [comment]
After some debugging i found out that my `GetBoundingBoxInLatLngWithMargin(int tileX, int tileY, int zoom)` was returning bounds that was way out of my data bounds. I realized i was applying a coordinate transformation to my bbox giving me a bbox in pseudo mercator yet my layer is in wgs84. I changed the `GetBoundingBoxInLatLngWithMargin `  to:

```
Envelope bbox = new Envelope();
bbox.ExpandToInclude(ll1.X, ll1.Y );
bbox.ExpandToInclude(ll2.X, ll2.Y);
return bbox;
```
Everything is now working fine. Thanks @FObermaier , i don't think would have found out that am missing call to Map.ZoomToExtents() without your help.
